MdeModulePkg/Core/Pei/Memory/MemoryServices.c

```diff
@@ -650,6 +652,30 @@ PeiAllocatePages (
     FreeMemoryBottom = &(Hob.HandoffInformationTable->EfiFreeMemoryBottom);
   }
 
+  // MU_CHANGE START
+  MemBucketHob = GetFirstGuidHob (&gMemoryBucketInformationGuid);
+  SyncMemoryBuckets (MemBucketHob);
+
+  // Check if we're using the memory buckets
+  if (IsRuntimeType (MemoryType)) {
+    if (IsRuntimeMemoryInitialized ()) {
+      *FreeMemoryTop    = GetCurrentBucketTop (MemoryType);
+      *FreeMemoryBottom = GetCurrentBucketBottom (MemoryType);
+    } else {
+      InitializeMemoryBuckets (*FreeMemoryTop);
+      if (IsRuntimeType (MemoryType)) {
+        *FreeMemoryTop    = GetCurrentBucketTop (MemoryType);
+        *FreeMemoryBottom = GetCurrentBucketBottom (MemoryType);
+      }
+    }

```

Comment:
I think you can clean up/reduce logic here.

```suggestion
  // Check if we're using the memory buckets
  if (IsRuntimeType (MemoryType)) {
    if (!IsRuntimeMemoryInitialized ()) {
      InitializeMemoryBuckets (*FreeMemoryTop);
    }

    *FreeMemoryTop    = GetCurrentBucketTop (MemoryType);
    *FreeMemoryBottom = GetCurrentBucketBottom (MemoryType);
    }
  }
```

Reply:
You're right about the general logic.  There is a reason why I include an additional IsRuntimeType check which is to deny memory bucket operations if after initializing we can find that the platform doesn't want to use memory buckets.  That should be its own separate function though, so I'll make that adjustment.